### PR TITLE
Add VPC search store

### PR DESCRIPTION
**Pull request by zhengxiexie**

### CODE REVIEW

File: pkg/nsx/services/common/store.go [hunk 1]
```diff
@@ -152,3 +152,42 @@ func (service *Service) InitializeResourceStore(wg *sync.WaitGroup, fatalErrors
 		}
 	}
 }
+
+// InitializeVPCResourceStore is the method to query all the various VPC resources from nsx-t side and
+// save them to the store, we could use it to cache all the resources when process starts.
+func (service *Service) InitializeVPCResourceStore(wg *sync.WaitGroup, fatalErrors chan error, org string, project string, resourceTypeValue string, store Store) {
```
Comment (TaoZou1) [inline pkg/nsx/services/common/store.go:158]: could InitializeVPCResourceStore merged with InitializeResourceStore? if org/project == nil, call the service.NSXClient.QueryClient.List, else call the service.NSXClient.VPCQueryClient.List

File: pkg/nsx/services/common/compare.go [hunk 2]
```diff
@@ -41,12 +41,15 @@ func CompareResources(existing []Comparable, expected []Comparable) (changed []C
...
 			}
 		}
 		changed = append(changed, e)
 	}
 	for key, e := range existingMap {
 		if _, ok := expectedMap[key]; !ok {
+			log.V(1).Info("resource stale", "existing", e)
 			stale = append(stale, e)
 		}
 	}
```
Comment (TaoZou1) [inline pkg/nsx/services/common/compare.go:55]: could line51-53 move to line 47 
47 } else {
  Line 51-53
}

Comment (zhengxiexie) [reply pkg/nsx/services/common/compare.go:55]:
```
for key, e := range expectedMap
for key, e := range existingMap
```
The object iterated is different

File: pkg/nsx/services/common/store.go (hunk 1 above)
Comment (zhengxiexie) [reply pkg/nsx/services/common/store.go:125]: I absorb a common method in the latest version, take a look

**Review by TaoZou1 — DISMISSED** [review]

File: pkg/nsx/services/common/store.go [hunk 3]
```diff
@@ -131,12 +143,21 @@ func (service *Service) InitializeResourceStore(wg *sync.WaitGroup, fatalErrors
 	resourceParam := fmt.Sprintf("%s:%s", ResourceType, resourceTypeValue)
 	queryParam := resourceParam + " AND " + tagParam
 
+	if org != "" || project != "" {
+		// QueryClient.List() will escape the path, "path:" then will be "path%25%3A" instead of "path:3A",
+		//"path%25%3A" would fail to get response. Hack it here.
+		path := "\\/orgs\\/" + org + "\\/projects\\/" + project + "\\/*"
```
Comment (lxiaopei) [inline pkg/nsx/services/common/store.go:149]: search can under "infra" scope or "org" or "project" scope?
If only org provided, that means search under "org" scope? then we do not need to add /projects/<project-id>

File: pkg/nsx/services/common/store.go [hunk 4]
```diff
@@ -123,6 +124,17 @@ func TransError(err error) error {
 // InitializeResourceStore is the method to query all the various resources from nsx-t side and
 // save them to the store, we could use it to cache all the resources when process starts.
 func (service *Service) InitializeResourceStore(wg *sync.WaitGroup, fatalErrors chan error, resourceTypeValue string, store Store) {
```
Comment (lxiaopei) [inline pkg/nsx/services/common/store.go:126]: this is to search all resources under "infra" right? Could we change to InitializeInfraResourceStore?

File: pkg/nsx/services/common/store.go (hunk 3 above)
Comment (zhengxiexie) [reply pkg/nsx/services/common/store.go:149]: Actually, InitializeVPCResourceStore is not used yet, we now only use InitializeResourceStore to query all resources, we can update it in the future if there are scenarios when only query under "org" or "org && project".

File: pkg/nsx/services/common/store.go (hunk 4 above)
Comment (zhengxiexie) [reply pkg/nsx/services/common/store.go:126]: It's a full-text search API according to SDK, it doesn't say it's under "infra", I believe it is a global search.

File: pkg/nsx/services/common/store.go [hunk 5]
```diff
@@ -131,12 +143,21 @@ func (service *Service) InitializeResourceStore(wg *sync.WaitGroup, fatalErrors
 	resourceParam := fmt.Sprintf("%s:%s", ResourceType, resourceTypeValue)
 	queryParam := resourceParam + " AND " + tagParam
 
+	if org != "" || project != "" {
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/nsx/services/common/store.go:146]: https://github.com/vmware-tanzu/nsx-operator/pull/172#discussion_r1201701450

File: pkg/nsx/services/common/compare.go [hunk 6]
```diff
@@ -41,12 +41,15 @@ func CompareResources(existing []Comparable, expected []Comparable) (changed []C
 		if e2, ok := existingMap[key]; ok {
 			if isChanged := CompareResource(e2, e); !isChanged {
 				continue
+			} else {
+				log.V(1).Info("resource changed", "existing", e2, "expected", e)
```
Comment (dantingl) [inline pkg/nsx/services/common/compare.go:45]: I guess `e2` is existing resource, `e` is expected one. The naming is confused..

Comment (zhengxiexie) [reply pkg/nsx/services/common/compare.go]: OK, updated.

**Review by dantingl — APPROVED** [review]

**Pull request merged**